fsspec/implementations/reference.py

```diff
@@ -119,63 +92,47 @@ def open_refs(path):
+            self.url = self.root + "/{field}/refs.{record}.parq"
+        else:
+            self.url = "{field}/refs.{record}.parq"
+
     def listdir(self, basename=True):
-        listing = self.fs.ls(self.root)
+        if self.dirs is None:
+            dirs = [p.split("/", 1)[0] for p in self.zmetadata]
+            self.dirs = set(sorted(p for p in dirs if p and not p.startswith(".")))
+        listing = self.dirs
         if basename:
             listing = [os.path.basename(path) for path in listing]
         return listing
 
-    def join(self, *args):
-        return self.fs.sep.join(args)
-
-    @property
-    def zmetadata(self):
-        if ".zmetadata" not in self._items:
-            self._get_and_cache_metadata()
-        return self._items[".zmetadata"]
-
-    @property
-    def record_size(self):
-        return self.zmetadata["record_size"]
-
     def _load_one_key(self, key):
-        if "/" not in key:
-            if key in self.zmetadata["metadata"]:
-                return self._get_and_cache_metadata(key)
-            if key not in self.listdir():
-                raise KeyError
-            return self._get_and_cache_metadata(key)
-        else:
-            field, sub_key = key.split("/")
-            if sub_key.startswith("."):
-                # zarr metadata keys are always cached
-                return self._get_and_cache_metadata(key)
-            # Chunk keys can be loaded from row group and cached in LRU cache
-            record, ri, chunk_size = self._key_to_record(key)
-            if chunk_size == 0:
-                return b""
-            pf_path = self.join(self.root, field, f"refs.{record}.parq")
-            refs = self.open_refs(pf_path)
-            columns = ["path", "offset", "size", "raw"]
-            selection = [refs[c][ri] if c in refs else None for c in columns]
-            raw = selection[-1]
-            if raw is not None:
-                return raw
-            data = selection[:-1]
-            if not isinstance(data[0], str):
-                data[0] = ""
-            return data
-
-    def _get_and_cache_metadata(self, key=None):
-        if key is None or ".zmetadata" not in self._items:
-            with self.fs.open(self.join(self.root, ".zmetadata"), "rb") as f:
-                self._items[".zmetadata"] = json.load(f)
-        if key is None:
-            return
-        if key == ".zmetadata":
-            # consolidated metadata JSON needs to be encoded as string for zarr
-            return json.dumps(self.zmetadata)
-        return self.zmetadata["metadata"][key]
+        if key in self._items:
+            return self._items[key]
+        elif key in self.zmetadata:
+            return json.dumps(self.zmetadata[key]).encode()
+        field, sub_key = key.split("/")
+        # Chunk keys can be loaded from row group and cached in LRU cache
+        record, ri, chunk_size = self._key_to_record(key)
+        if chunk_size == 0:
+            # raise KeyError ?
```

Comment:
This check was added because some refs I have tested would store variables consisting of a single empty string as a 0 length zarr array instead of as a global attribute.

Reply:
A length-zero array has no chunk file but does have attributes?
So not a KeyError, please ignore.

Reply:
Yep!